src/main/kotlin/org/rust/lang/core/type/TypeInference.kt

```diff
@@ -1,32 +1,47 @@
 package org.rust.lang.core.type
 
-import org.rust.lang.core.psi.*
-import org.rust.lang.core.psi.util.parentOfType
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.PsiModificationTracker
+import org.rust.lang.core.psi.RustExprElement
+import org.rust.lang.core.psi.RustTypeElement
+import org.rust.lang.core.psi.visitors.RustTypificationVisitor
+import org.rust.lang.core.type.unresolved.RustUnresolvedType
+import org.rust.lang.core.type.visitors.RustTypeResolvingVisitor
 import org.rust.lang.utils.psiCached
 
-val RustExprElement.inferredType: RustResolvedType by psiCached {
-    when (this) {
-        is RustPathExprElement -> {
-            val target = path.reference.resolve()
-            when (target) {
-                is RustSelfArgumentElement -> target.parentOfType<RustImplItemElement>()?.type?.resolvedType ?: RustUnknownType
-                else -> RustUnknownType
-            }
-        }
-        else -> RustUnknownType
+val RustExprElement.type: RustUnresolvedType by psiCached {
```

Comment:
Ouch, it was indeed a terrible bug. But I wonder if we can use some Kotlin features to still get the simple API for this one. Perhaps an inline function can help?

```
/**
 * Cached value invalidated on any PSI modification
 */
inline fun<E : PsiElement, T> psiCached(noinline block: E.() -> T): PsiCacheDelegate<E, T> {
    val provider = {element: E -> CachedValueProvider {
        CachedValueProvider.Result.create(element.block(), PsiModificationTracker.MODIFICATION_COUNT) }
    }
    return PsiCacheDelegate(provider)
}


class PsiCacheDelegate<E : PsiElement, T>(val provider: (E) -> CachedValueProvider<T>) {
    operator fun getValue(element: E, property: KProperty<*>): T {
        return CachedValuesManager.getCachedValue(element, provider(element))
    }
}
```